python/lsst/daf/butler/butler.py

```diff
@@ -61,17 +62,54 @@ class Butler:
     ----------
     config : `Config`
         Configuration.
+    collection : `str` or `None`
+        Collection to use for all input lookups, overriding config['collection']
+        if provided.
+    run : `str`, `Run`, or `None`
+        Collection associated with the `Run` to use for outputs, overriding
+        config['run'].  If a `Run` associated with the given Collection does
+        not exist, it will be created.  If "collection" is None, this
+        collection will be used for input lookups as well; if not, it must have
+        the same value as "run".
+
+    Raises
+    ------
+    ValueError
+        Raised if neither 'collection' nor 'run' are provided by argument or
+        config, or if both are provided and are inconsistent.
     """
 
-    def __init__(self, config):
+    def __init__(self, config, collection=None, run=None):
         self.config = ButlerConfig(config)
         self.registry = Registry.fromConfig(self.config)
         self.datastore = Datastore.fromConfig(self.config, self.registry)
         self.storageClasses = StorageClassFactory()
         self.storageClasses.addFromConfig(self.config)
-        self.run = self.registry.getRun(collection=self.config['run'])
-        if self.run is None:
-            self.run = self.registry.makeRun(self.config['run'])
+        if run is None:
+            run = self.config.get("run", None)
+        else:
+            if isinstance(run, Run):
+                self.run = run
+                run = self.run.collection
+            else:
+                self.run = None
+            # if run *arg* is not None and collection arg is, use run for collecion.
+            if collection is None:
+                collection = run
+        # n.b. at this point, 'run' is a str or None; 'self.run' is a Run or None
+        if collection is None:  # didn't get a collection from collection or run *args*
+            collection = self.config.get("collection", None)
+            if collection is None:  # didn't get a collection from config['collection']
+                collection = run    # get collection from run found in config
+        if collection is None:
+            raise ValueError("No run or collection provided.")
+        if run is not None and collection != run:
```

Comment:
I'll give it a try, but I'll want to see how it works before committing; I'd like to leave the argument as just `run` because it actually can be a `Run`.